File: loom/Bench1.java
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousFileChannel;
import java.nio.channels.CompletionHandler;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

public class Bench1 {
    static private Thread[] threadPool = null;
    static private final int NUM_HELLO = 1000000;
    static private final String message = "hello ";

    public static void task(int numCoroutines, AsynchronousFileChannel channel) throws InterruptedException {
        AtomicInteger inc = new AtomicInteger();
        AtomicLong startPos = new AtomicLong(0L);
        for (int i = 0; i < numCoroutines; i++) {
            threadPool[i] = Thread.startVirtualThread(() -> {
                while (inc.get() < NUM_HELLO) {
                    inc.addAndGet(1);
                    var pos = startPos.getAndAdd(message.length());
                    write(channel, pos, ByteBuffer.wrap(message.getBytes()));
                }
            });
        }

        for (Thread th : threadPool) {
            th.join();
        }
    }

    private static void write(AsynchronousFileChannel ch, long pos, ByteBuffer buf) {
        ch.write(buf, pos, "nothing", new CompletionHandler<>() {
            @Override
            public void completed(Integer result, String attachment) {
                //Todo
            }

            @Override
            public void failed(Throwable e, String attachment) {
                assert false;
            }
        });
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        var numCoroutines = Integer.parseInt(args[0]);
        var path = Path.of(args[1]);

        var ch = AsynchronousFileChannel.open(path, StandardOpenOption.WRITE, StandardOpenOption.CREATE);
        threadPool = new Thread[numCoroutines];

        long startTime = System.currentTimeMillis();
        task(numCoroutines, ch);
        long endTime = System.currentTimeMillis();

        System.out.println("Execution time: " + (endTime - startTime) + "ms.");
    }
}

File: kotlin/bench1/src/main/kotlin/org/coroutines/benchmark/Bench1.kt
package org.coroutines.benchmark

import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.nio.ByteBuffer
import java.nio.channels.AsynchronousFileChannel
import java.nio.channels.CompletionHandler
import java.nio.file.Paths
import java.nio.file.StandardOpenOption
import java.util.concurrent.atomic.AtomicInteger
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

val MESSAGE = "hello "
val NUM_HELLO = 1_000_000

fun task(channel: AsynchronousFileChannel, numCoroutines: Long) {
    var startPos = 0L
    var inc = 0
    for (i in 1L..numCoroutines) {
        runBlocking {
            launch {
                while (inc < NUM_HELLO) {
                    channel.WriteString(MESSAGE, startPos)
                    startPos += MESSAGE.length
                    inc += 1
                }
            }
        }
    }
}

suspend fun AsynchronousFileChannel.WriteBuffer(buf: ByteBuffer, pos: Long = 0): Int =
        suspendCoroutine { cont ->
            write(buf, pos, Unit, object : CompletionHandler<Int, Unit> {
                override fun completed(bytesWrote: Int, attachment: Unit) {
                    cont.resume(bytesWrote)
                }

                override fun failed(exception: Throwable, attachment: Unit) {
                    cont.resumeWithException(exception)
                }
            })
        }

suspend fun AsynchronousFileChannel.WriteString(string: String, pos: Long = 0): Int =
        WriteBuffer(ByteBuffer.wrap(string.toByteArray()), pos)

fun main(args: Array<String>) {
    val numCoroutines = args[0].toLong()
    val file = Paths.get(args[1])

    val ch = AsynchronousFileChannel.open(file, StandardOpenOption.WRITE, StandardOpenOption.CREATE)

    val startTime = System.currentTimeMillis()
    task(ch, numCoroutines)
    val endTime = System.currentTimeMillis()

    println("Execution time: " + (endTime - startTime) + "ms.")
    ch.close()
}